sycl/doc/design/MappingHostAddressesToDeviceEntities.md

```diff
@@ -0,0 +1,344 @@
+} // namespace inner
+
+// ...
+```
+
+The integration footer generated by the compiler contains the definition of the
+constructor, which calls a function in the DPC++ runtime, which registers
+needed mappings:
+
+```
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  __register_uniquely_identifiable_object(
+    &::Foo,
+    /* same string returned from __builtin_sycl_unique_stable_id(::Foo) */,
+    "device_global");
+  __register_uniquely_identifiable_object(
+    &::inner::Bar,
+    /* same string returned from __builtin_sycl_unique_stable_id(::inner::Bar) */,
+    "device_global");
+}
+
+} // namespace (unnamed)
+} // namespace sycl::detail
+```
+
+### Handling shadowed variables
+
+The example above shows a simple case where the user's device global variables
+can all be uniquely referenced via fully qualified lookup (e.g.
+`::inner::Bar`).  However, it is possible for users to construct applications
+where this is not the case, for example:
+
+```
+sycl::device_global<int> FuBar;
+namespace {
+  sycl::device_global<int> FuBar;
+}
+```
+
+In this example, the `FuBar` variable in the global namespace shadows a
+variable with the same name in the unnamed namespace.  The integration footer
+can reference the variable in the global namespace as `::FuBar`, but there is
+no way to reference the variable in the unnamed namespace using fully qualified
+lookup.
+
+Such programs are still legal, though.  The integration footer can support
+cases like this by defining a shim function that returns a reference to the
+shadowed device global:
+
+```
+namespace {
+namespace __sycl_detail {
+
+static constexpr decltype(FuBar) &__shim_1() {
+  return FuBar;   // References 'FuBar' in the unnamed namespace
+}
+
+} // namespace __sycl_detail
+} // namespace (unnamed)
+
+namespace sycl::detail {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  __register_uniquely_identifiable_object(
+    &::FuBar,
+    /* same string returned from __builtin_sycl_unique_stable_id(::FuBar) */,
+    "device_global");
+  __register_uniquely_identifiable_object(
+    &::__sycl_detail::__shim_1(),
+    /* same string returned from __builtin_sycl_unique_stable_id(::(unnamed)::FuBar) */,
+    "device_global");
+}
+
+} // namespace sycl::detail
+```
+
+The `__shim_1()` function is defined in the same namespace as the second
+`FuBar` device global, so it can reference the variable through unqualified
+name lookup.  Furthermore, the name of the shim function is globally unique, so
+it is guaranteed not to be shadowed by any other name in the translation unit.
+This problem with variable shadowing is also a problem for the integration
+footer we use for specialization constants.  See the [specialization constant
+design document][5] for more details on this topic.
+
+[5]: <SpecializationConstants.md>
+
+## Custom host compiler approach
```

Comment:
This is when we use DPCPP host compiler right? 'Custom' host compiler is confusing here because sometimes that is how we refer to 3rd party host compilers

Reply:
Done in e963911a9455e5508174ae38d51a313d68182d47